### PR TITLE
fix: remove reentrancy in ex13

### PR DESCRIPTION
**Pull request by Tristan22400**

### CODE REVIEW

**Comment by makcandrov** [comment]
I think a re-entrancy is still possible here

**Comment by makcandrov** [comment]
I'd have fixed it like that

```
function distribute(uint256 _count) public {
        require(currentStage == Stage.WinnersFound, "Winners must have been found");
        while (
            cursorDistribute < winners.length
            && _count != 0
        ) {
            // Send ether to the winners. Do not block if one of the account cannot receive ETH.
            winners[cursorDistribute++].call{value: totalBalance / (winners.length - numberOfLosers)}("");
            _count--;
        }
        if (cursorDistribute == winners.length - 1) currentStage = Stage.Distributed;
    }
```

though I don't know how to get rid of the warning

**Comment by Tristan22400** [comment]
> I think a re-entrancy is still possible here

Can you explain it? I do not see how it is possible.

**Comment by Tristan22400** [comment]
> I'd have fixed it like that
> 
> ```
> function distribute(uint256 _count) public {
>         require(currentStage == Stage.WinnersFound, "Winners must have been found");
>         while (
>             cursorDistribute < winners.length
>             && _count != 0
>         ) {
>             // Send ether to the winners. Do not block if one of the account cannot receive ETH.
>             winners[cursorDistribute++].call{value: totalBalance / (winners.length - numberOfLosers)}("");
>             _count--;
>         }
>         if (cursorDistribute == winners.length - 1) currentStage = Stage.Distributed;
>     }
> ```
> 
> though I don't know how to get rid of the warning

I agree with you. I do not know how to resolve it.

File: src/Hack.sol [hunk 1]
```diff
@@ -602,13 +602,16 @@ contract GuessTheAverage {
...
+
+        while (cursorDistribute < winners.length && _count != 0) {
+            // Send ether to the winners. Do not block if one of the account cannot receive ETH.
+            winners[cursorDistribute++].call{value: totalBalance / (winners.length - numberOfLosers)}("");
+            _count--;
         }
+
+        if (cursorDistribute == winners.length - 1) currentStage = Stage.Distributed;
+
         // Update the cursor in case we haven't finished going through the list.
```
Comment (MerlinEgalite) [inline src/Hack.sol:614]: to remove no?

File: src/Hack.sol [hunk 2]
```diff
@@ -602,13 +602,16 @@ contract GuessTheAverage {
...
     function distribute(uint256 _count) public {
         require(currentStage == Stage.WinnersFound, "Winners must have been found");
-        for (uint256 i = cursorDistribute; i < winners.length && (_count == 0 || i < cursorDistribute + _count); i++) {
-            // Send ether to the winners, use send not to block, if one of the account cannot receive ETH.
-            winners[i].call{value: totalBalance / (winners.length - numberOfLosers)}("");
-            if (i == winners.length - 1) currentStage = Stage.Distributed;
+
+        while (cursorDistribute < winners.length && _count != 0) {
+            // Send ether to the winners. Do not block if one of the account cannot receive ETH.
+            winners[cursorDistribute++].call{value: totalBalance / (winners.length - numberOfLosers)}("");
```
Comment (MerlinEgalite) [inline src/Hack.sol:608]: that's not ideal in term of storage writing, could be better if `cursorDistribute` is stored in a temp variable?

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]

File: src/Hack.sol (hunk 2 above)
Comment (MerlinEgalite) [reply src/Hack.sol:608]: ok haha let's go

**Review by MerlinEgalite — APPROVED** [review]

**Pull request merged**